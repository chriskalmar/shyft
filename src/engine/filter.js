import { isEntity } from './entity/Entity';

import {
  passOrThrow,
  isMap,
  isArray,
  mapOverProperties,
  isFunction,
} from './util';
import { isObjectDataType } from './datatype/ObjectDataType';

const logicFilters = [ '$and', '$or' ];
const deepFilter = '$filter';
const preFilter = '$pre_filter';
const noResultFilter = '$noResult';

export const validateFilterLevel = (filters, attributes, path, storageType) => {
  const ret = {};

  passOrThrow(!path || isArray(path, true), () => {
    return 'optional path in validateFilterLevel() needs to be an array';
  });

  const errorLocation = path ? ` at '${path.join('.')}'` : '';

  passOrThrow(
    isMap(filters),
    () => `filter${errorLocation} needs to be an object of filter criteria`,
  );

  passOrThrow(isMap(attributes, true), () => {
    return 'validateFilterLevel() expects an attribute map';
  });

  mapOverProperties(filters, (value, filter) => {
    if (logicFilters.includes(filter)) {
      const newPath = path ? path.slice() : [];

      newPath.push(filter);
      value.map(newFilter =>
        validateFilterLevel(newFilter, attributes, path, storageType),
      );

      return;
    }

    if (filter.indexOf(deepFilter) === 0) {
      const newPath = path ? path.slice() : [];

      newPath.push(filter);

      const attributeName = filter.replace(`${deepFilter}.`, '');
      const targetEntity = attributes[attributeName].type;

      validateFilterLevel(
        value,
        targetEntity.getAttributes(),
        path,
        storageType,
      );

      return;
    }

    const attributeName = filter;
    const attribute = attributes[attributeName];

    passOrThrow(attribute, () => {
      return `Unknown attribute name '${attributeName}' used in filter${errorLocation}`;
    });

    passOrThrow(!attribute.mutationInput, () => {
      return `Mutation input attribute '${attributeName}' not allowed in filter${errorLocation}`;
    });

    if (isMap(value)) {
      let storageDataType;

      if (isEntity(attribute.type)) {
        const primaryAttribute = attribute.type.getPrimaryAttribute();
        storageDataType = storageType.convertToStorageDataType(
          primaryAttribute.type,
        );
      }
      else {
        storageDataType = storageType.convertToStorageDataType(attribute.type);
      }

      const operators = Object.keys(value);

      operators.map(operator => {
        if (operator === noResultFilter) {
          return;
        }

        if (operator === preFilter && attribute.primary) {
          return;
        }

        const operatorCapabilityName = operator.replace('$', '');
        passOrThrow(
          storageDataType.capabilities.indexOf(operatorCapabilityName) >= 0,
          () => {
            return `Unknown or incompatible operator '${operator}' used on '${attributeName}' in filter${errorLocation}`;
          },
        );
      });
    }
  });

  return ret;
};

export const processFilter = (entity, args, storageType) => {
  const { filter } = args;

  if (!filter) {
    return {};
  }

  validateFilterLevel(filter, entity.getAttributes(), null, storageType);

  const where = {
    ...filter,
  };

  return where;
};

export const convertFilterLevel = (filterShaper, filterLevel) => {
  const converted = filterShaper(filterLevel);
  const filterLevelKeys = Object.keys(converted);
  const ret = {};

  filterLevelKeys.map(key => {
    const filter = filterLevel[key];

    if (filter) {
      if (isMap(filter)) {
        ret[key] = convertFilterLevel(filterShaper, filter);
      }
      else if (logicFilters.includes(key)) {
        ret[key] = filter.map(item => convertFilterLevel(filterShaper, item));
      }
      else {
        ret[key] = converted[key];
      }
    }
  });

  filterLevelKeys.map(key => {
    if (
      typeof converted[key] !== 'undefined' &&
      typeof ret[key] === 'undefined'
    ) {
      ret[key] = converted[key];
    }
  });

  return ret;
};

const isPreFilter = preFilterDefinition => {
  if (isMap(preFilterDefinition)) {
    if (isFunction(preFilterDefinition.resolve)) {
      if (
        !preFilterDefinition.attributes ||
        isObjectDataType(preFilterDefinition.attributes)
      ) {
        return true;
      }
    }
  }

  return false;
};

export const processPreFilters = (entity, preFilters) => {
  passOrThrow(
    isMap(preFilters),
    () =>
      `Entity '${
        entity.name
      }' preFilters definition needs to be a map with individual preFilters`,
  );

  Object.keys(preFilters).map(preFilterName => {
    const preFilterDefinition = preFilters[preFilterName];

    passOrThrow(
      isPreFilter(preFilterDefinition),
      () =>
        `Invalid preFilter definition '${preFilterName}' for entity '${
          entity.name
        }'`,
    );
  });

  return preFilters;
};
