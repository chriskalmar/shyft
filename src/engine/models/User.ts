import { Entity } from '../entity/Entity';
import { DataTypeString } from '../datatype/dataTypes';
import { Index, INDEX_UNIQUE } from '../index/Index';

import { Language } from './Language';

export const User = new Entity({
  name: 'User',
  description: 'A user',

  // isUserTable: true,
  // sequenceGenerator: 'timebased',

  indexes: [
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'username' ],
    }),
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'email' ],
    }),
  ],

  attributes: {
    username: {
      type: DataTypeString,
      description: 'Username',
      required: true,
    },

    firstName: {
      type: DataTypeString,
      description: 'First name',
    },

    lastName: {
      type: DataTypeString,
      description: 'Last name',
    },

    language: {
      type: Language,
      description: 'User language',
      required: true,
    },

    email: {
      type: DataTypeString,
      description: 'Email address',
      required: true,
    },

    password: {
      type: DataTypeString,
      description: 'User password',
      required: true,
    },
  },
});
