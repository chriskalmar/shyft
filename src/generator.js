import StorageTypePostgres from './StorageTypePostgres';
import _ from 'lodash';
import toposort from 'toposort';
import { generateIndexName } from './util';

import {
  isEntity,
  INDEX_UNIQUE,
  INDEX_GENERIC,
  mapOverProperties,
  isComplexDataType,
  isConfiguration,
} from 'shyft';

import { shaper } from 'json-shaper';

import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  Index,
  ManyToOne,
  JoinColumn,
  createConnection,
} from 'typeorm';

import { isStoragePostgresConfiguration } from './StoragePostgresConfiguration';

const filterOperatorMap = {
  $ne: '$ne',

  $in: '$in',
  $notIn: '$not_in',

  $lt: '$lt',
  $lte: '$lte',

  $gt: '$gt',
  $gte: '$gte',

  $startsWith: '$starts_with',
  $contains: '$contains',
  $endsWith: '$ends_with',

  $notStartsWith: '$not_starts_with',
  $notContains: '$not_contains',
  $notEndsWith: '$not_ends_with',

  $noResult: '$noResult',

  $not: '$not',
  $or: '$or',
  $and: '$and',
};

export const loadModels = configuration => {
  if (!isConfiguration(configuration)) {
    throw new Error('Invalid configuration object provided to loadModels()');
  }

  const schema = configuration.getSchema();
  const storageConfiguration = configuration.getStorageConfiguration();

  if (!isStoragePostgresConfiguration(storageConfiguration)) {
    throw new Error(
      'Invalid storage configuration object provided to loadModels()',
    );
  }

  StorageTypePostgres.setStorageConfiguration(storageConfiguration);

  const modelRegistry = {};

  _.map(schema.getEntities(), (entity, entityName) => {
    if (String(entity.storageType) !== String(StorageTypePostgres)) {
      return;
    }

    const attributes = {};
    const references = [];
    const dataShaperMap = {};
    const filterShaperMap = {};

    const Skeleton = () => {};

    Object.defineProperty(Skeleton, 'name', {
      value: entityName,
    });

    _.map(entity.getAttributes(), attribute => {
      // skip for computed values
      if (attribute.resolve) {
        return;
      }

      // skip for mutation inputs
      if (attribute.mutationInput) {
        return;
      }

      const attributeName = attribute.name;
      const storageAttributeName =
        attribute.meta && attribute.meta.storageAttributeName
          ? attribute.meta.storageAttributeName
          : _.snakeCase(attribute.name);

      dataShaperMap[attributeName] = storageAttributeName;
      filterShaperMap[attributeName] = attributeName;

      let storageDataType;

      // it's a reference
      if (isEntity(attribute.type)) {
        const primaryAttribute = attribute.type.getPrimaryAttribute();
        storageDataType = StorageTypePostgres.convertToStorageDataType(
          primaryAttribute.type,
        );

        const targetEntityName = attribute.type.name;

        if (attribute.targetAttributesMap) {
          // sequelize doesn't support composite foreign keys yet (so this has no impact):
          // https://github.com/sequelize/sequelize/issues/311
          mapOverProperties(
            attribute.targetAttributesMap,
            (targetAttribute, sourceAttributeName) => {
              references.push({
                sourceAttributeName,
                targetAttributeName: targetAttribute.name,
                targetEntityName,
              });
            },
          );
        }
        else {
          references.push({
            sourceAttributeName: attributeName,
            targetAttributeName: primaryAttribute.name,
            targetEntityName,
          });
        }

        ManyToOne(() => modelRegistry[targetEntityName].model, {
          onDelete: 'CASCADE',
        })(Skeleton.prototype, storageAttributeName);

        JoinColumn({
          name: storageAttributeName,
          cascadeAll: true,
        })(Skeleton.prototype, storageAttributeName);
      }
      // it's a regular attribute
      else {
        storageDataType = StorageTypePostgres.convertToStorageDataType(
          attribute.type,
        );
      }

      attributes[attributeName] = {
        name: storageAttributeName,
        type: storageDataType.nativeDataType,
        primary: attribute.primary,
        nullable: !attribute.required,
      };

      if (attribute.primary) {
        const primaryGenerator =
          storageDataType.name === 'StorageDataTypeUUID' ? 'uuid' : 'increment';

        PrimaryGeneratedColumn(primaryGenerator, attributes[attributeName])(
          Skeleton.prototype,
          attributeName,
        );
      }
      else {
        Column(attributes[attributeName])(Skeleton.prototype, attributeName);
      }
    });

    const constraints = {
      unique: {},
    };

    if (entity.indexes) {
      entity.indexes.map(index => {
        if (index.type === INDEX_UNIQUE || index.type === INDEX_GENERIC) {
          const indexAttributes = index.attributes.map(_.snakeCase);
          const indexName = generateIndexName(
            _.snakeCase(entityName),
            indexAttributes,
            'key',
          );

          const unique = index.type === INDEX_UNIQUE;
          Index(indexName, index.attributes, { unique })(Skeleton);

          constraints.unique[indexName] = {
            attributes: index.attributes,
          };
        }
      });
    }

    const storageTableName = _.snakeCase(entityName);
    entity.storageTableName = storageTableName;

    Entity({ name: storageTableName })(Skeleton);

    modelRegistry[entityName] = {
      model: Skeleton,
      storageTableName,
      dataShaperMap,
      reverseDataShaperMap: _.invert(dataShaperMap),
      dataShaper: shaper(dataShaperMap),
      reverseDataShaper: shaper(_.invert(dataShaperMap)),
      filterShaper: shaper(_.extend(filterShaperMap, filterOperatorMap)),
      references,
      constraints,
    };
  });

  return modelRegistry;
};

export const generateMockData = async configuration => {
  if (!isConfiguration(configuration)) {
    throw new Error(
      'Invalid configuration object provided to generateMockData()',
    );
  }

  const schema = configuration.getSchema();
  const languages = configuration.getLanguages();
  const storageConfiguration = configuration.getStorageConfiguration();
  const storageInstance = storageConfiguration.getStorageInstance();
  const modelRegistry = storageConfiguration.getStorageModels();

  const edges = [];
  const entities = schema.getEntities();
  const mockInstancesMemory = {};

  _.map(entities, entity => {
    _.map(entity.getAttributes(), attribute => {
      if (isEntity(attribute.type)) {
        // skip cyclic dependencies
        if (entity.name === attribute.type.name) {
          return;
        }

        edges.push([ attribute.type.name, entity.name ]);
      }
    });
  });

  const dependencySortedEntityNames = toposort(edges);

  // add independent entities
  _.map(entities, entity => {
    if (dependencySortedEntityNames.indexOf(entity.name) === -1) {
      dependencySortedEntityNames.push(entity.name);
    }
  });

  for (let e = 0; e < dependencySortedEntityNames.length; e++) {
    const entityName = dependencySortedEntityNames[e];
    const entity = entities[entityName];

    if (String(entity.storageType) !== String(StorageTypePostgres)) {
      return;
    }

    try {
      const mockItemCount =
        entity.meta && entity.meta.mockItemsCount >= 0
          ? entity.meta.mockItemsCount
          : _.random(10, 100);

      for (let i = 0; i < mockItemCount; i++) {
        await generateItem(
          storageInstance,
          entity,
          mockInstancesMemory,
          modelRegistry[entityName],
          languages,
        );
      }
    }
    catch (err) {
      console.error(err);
    }
  }
};

const getExistingData = async (
  storageInstance,
  mockInstancesMemory,
  entity,
) => {
  const { name: entityName, storageTableName } = entity;

  const manager = storageInstance.manager;
  const result = await manager
    .createQueryBuilder(storageTableName, 't')
    .select([ 't.id' ])
    .getMany();

  if (result.length) {
    mockInstancesMemory[entityName] = result.map(({ id }) => id);
  }
  else {
    mockInstancesMemory[entityName] = [];
  }
};

async function generateItem(
  storageInstance,
  entity,
  mockInstancesMemory,
  model,
  languages,
) {
  const { name: entityName, storageTableName } = entity;
  const item = {};

  await Promise.all(
    _.map(
      entity.getAttributes(),
      async ({ type, primary, required, mock }, name) => {
        if (primary) {
          return;
        }

        if (!required) {
          if (!entity.meta || (entity.meta && !entity.meta.mockNoNulls)) {
            if (Math.random() > 0.5) {
              return;
            }
          }
        }

        if (isEntity(type)) {
          if (!mockInstancesMemory[type.name]) {
            await getExistingData(storageInstance, mockInstancesMemory, type);
          }

          item[name] = _.sample(mockInstancesMemory[type.name]);
        }
        if (isComplexDataType(type)) {
          // TODO: generate mocks based on complex types
          item[name] = {};
        }
        else {
          const mockDataGenerator = mock || type.mock;

          if (mockDataGenerator) {
            item[name] = mockDataGenerator(entity, name, model, languages);
          }
        }
      },
    ),
  );

  const manager = storageInstance.manager;
  let instance;

  const result = await manager
    .createQueryBuilder()
    .insert()
    .into(storageTableName)
    .values(item)
    .returning('*')
    .execute();

  if (result.raw.length) {
    instance = result.raw[0];

    const { name: primaryAttributeName } = entity.getPrimaryAttribute();

    mockInstancesMemory[entityName] = mockInstancesMemory[entityName] || [];
    mockInstancesMemory[entityName].push(instance[primaryAttributeName]);
  }
}

export const installStorageScripts = async (
  configuration,
  synchronize = false,
) => {
  const storageConfiguration = configuration.getStorageConfiguration();
  const languages = configuration.getLanguages();
  const storageInstance = storageConfiguration.getStorageInstance();
  const manager = storageInstance.manager;
  const modelFeatures = {};
  const schema = configuration.getSchema();

  if (languages.length > 1) {
    modelFeatures.language = true;
  }

  _.forEach(schema.getEntities(), entity => {
    if (entity.getI18nAttributeNames()) {
      modelFeatures.i18n = true;
    }

    if (entity.hasStates()) {
      modelFeatures.state = true;
    }
  });

  if (modelFeatures.state) {
    await manager.query(
      storageConfiguration.generateGetStateIdFunction(configuration),
    );
    await manager.query(
      storageConfiguration.generateGetStateIdsFunction(configuration),
    );
    await manager.query(
      storageConfiguration.generateGetStateMapFunction(configuration),
    );
    await manager.query(
      storageConfiguration.generateGetStateNameFunction(configuration),
    );
  }

  if (modelFeatures.i18n) {
    await manager.query(
      storageConfiguration.generateGetAttributeTranslationFunction(
        configuration,
      ),
    );
    await manager.query(
      storageConfiguration.generateGetAttributeTranslationsFunction(
        configuration,
      ),
    );
    await manager.query(
      storageConfiguration.generateMergeTranslationsFunction(configuration),
    );

    if (synchronize) {
      await manager.query(
        storageConfiguration.createI18nIndices(configuration, manager),
      );
    }
  }
};

let connection;

export const connectStorage = async (configuration, synchronize = false) => {
  const storageConfiguration = configuration.getStorageConfiguration();
  const connectionConfig = storageConfiguration.getConnectionConfig();

  const modelRegistry = loadModels(configuration);

  const entities = Object.keys(modelRegistry).map(entityName => {
    return modelRegistry[entityName].model;
  });

  connection = await createConnection({
    ...connectionConfig,
    type: 'postgres',
    synchronize,
    dropSchema: synchronize,
    entities,
  });

  storageConfiguration.setStorageInstance(connection);
  storageConfiguration.setStorageModels(modelRegistry);

  // remember logger settings and disable logger
  const loggerOptions = connection.logger.options;
  connection.logger.options = false;

  await installStorageScripts(configuration, synchronize);

  // bring back logger options
  connection.logger.options = loggerOptions;
};

export const disconnectStorage = async () => {
  await connection.close();
};

export const getConnection = () => connection;
