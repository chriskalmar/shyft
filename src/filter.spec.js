import { createConnection } from 'typeorm';
import { buildWhereQuery, purifyFilter } from './filter';
import { StorageTypePostgres } from './StorageTypePostgres';
import { loadModels } from './generator';
import { Server } from '../test/models/Server';
import StoragePostgresConfiguration from './StoragePostgresConfiguration';
import { Schema, Configuration } from 'shyft';

let connection;
let modelRegistry;

beforeAll(async () => {
  const schema = new Schema({
    defaultStorageType: StorageTypePostgres,

    entities: [Server],
  });

  const storageConfiguration = new StoragePostgresConfiguration();

  const configuration = new Configuration({
    schema,
    storageConfiguration,
  });

  modelRegistry = loadModels(configuration);

  const entities = Object.keys(modelRegistry).map(entityName => {
    return modelRegistry[entityName].model;
  });

  connection = await createConnection({
    database: ':memory:',
    type: 'sqlite',
    entities,
  });
});

describe('filter', () => {
  describe('purifyFilter', () => {
    it('should clean empty filter blocks', () => {
      const filter = {
        $and: [
          {
            username: {
              $contains: 'bot',
              $startsWith: '_system_%',
            },
            language: {
              $in: [],
              $lt: 5,
              $gte: 1,
            },
            $or: [
              {
                'LookupEntity.role': 1,
              },
              {},
              {
                accessLevel: 1,
              },
            ],
          },
          {
            $and: [{}, {}],
          },
          {
            $or: [
              {},
              {
                $and: [
                  {
                    username: 'hallo',
                  },
                ],
              },
            ],
          },
        ],
        $or: [
          {
            lastname: 'smith',
            firstname: {},
            age: null,
          },
          {},
        ],
      };

      const pureFilter = purifyFilter(filter);
      expect(pureFilter).toMatchSnapshot();
    });
  });

  describe('buildWhereQuery', () => {
    const entityNameServer = 'Server';
    const storageTableNameServer = 'server';

    it('empty filter', () => {
      const qBuilder = connection.createQueryBuilder(
        storageTableNameServer,
        entityNameServer,
      );
      const filter = null;

      buildWhereQuery(qBuilder, filter, entityNameServer, modelRegistry);
      const query = qBuilder.getQueryAndParameters();
      expect(query).toMatchSnapshot();
    });

    it('simple properties filter', () => {
      const qBuilder = connection.createQueryBuilder(
        storageTableNameServer,
        entityNameServer,
      );
      const filter = {
        ip: '127.0.0.1',
        name: 'test-server-1',
      };

      buildWhereQuery(qBuilder, filter, entityNameServer, modelRegistry);
      const query = qBuilder.getQueryAndParameters();
      expect(query).toMatchSnapshot();
    });

    it('accept null as value', () => {
      const qBuilder = connection.createQueryBuilder(
        storageTableNameServer,
        entityNameServer,
      );
      const filter = {
        ip: '127.0.0.1',
        name: null,
        clusterZone: {
          $ne: null,
        },
      };

      buildWhereQuery(qBuilder, filter, entityNameServer, modelRegistry);
      const query = qBuilder.getQueryAndParameters();
      expect(query).toMatchSnapshot();
    });

    it('logical AND filter', () => {
      const qBuilder = connection.createQueryBuilder(
        storageTableNameServer,
        entityNameServer,
      );
      const filter = {
        $and: [
          {
            ip: '127.0.0.1',
            name: 'test-server-1',
          },
          {
            clusterZone: 4,
          },
        ],
      };

      buildWhereQuery(qBuilder, filter, entityNameServer, modelRegistry);
      const query = qBuilder.getQueryAndParameters();
      expect(query).toMatchSnapshot();
    });

    it('logical OR filter', () => {
      const qBuilder = connection.createQueryBuilder(
        storageTableNameServer,
        entityNameServer,
      );
      const filter = {
        $or: [
          {
            ip: '127.0.0.1',
            name: 'test-server-1',
          },
          {
            ip: '127.0.0.1',
            name: 'test-server-2',
          },
        ],
      };

      buildWhereQuery(qBuilder, filter, entityNameServer, modelRegistry);
      const query = qBuilder.getQueryAndParameters();
      expect(query).toMatchSnapshot();
    });

    it('combined logical AND/OR filter', () => {
      const qBuilder1 = connection.createQueryBuilder(
        storageTableNameServer,
        entityNameServer,
      );

      const filter1 = {
        $and: [
          {
            ip: '127.0.0.1',
            $or: [
              {
                clusterZone: 1,
              },
              {},
              {
                clusterZone: 7,
              },
            ],
          },
          {
            name: 'test-server-1',
          },
        ],
      };

      buildWhereQuery(qBuilder1, filter1, entityNameServer, modelRegistry);
      const query1 = qBuilder1.getQueryAndParameters();
      expect(query1).toMatchSnapshot();

      const qBuilder2 = connection.createQueryBuilder(
        storageTableNameServer,
        entityNameServer,
      );

      const filter2 = {
        $or: [
          {
            ip: '127.0.0.1',
            $and: [
              {
                clusterZone: 1,
              },
              {},
              {
                name: 'test-server-110',
              },
            ],
          },
          {
            name: 'test-server-1',
          },
        ],
      };

      buildWhereQuery(qBuilder2, filter2, entityNameServer, modelRegistry);
      const query2 = qBuilder2.getQueryAndParameters();
      expect(query2).toMatchSnapshot();
    });

    it('reject if unknown operators are used', () => {
      const qBuilder = connection.createQueryBuilder(
        storageTableNameServer,
        entityNameServer,
      );

      const filter = {
        $or: [
          {
            name: {
              $unknownOp: 123,
            },
          },
        ],
      };

      const fn = () => {
        buildWhereQuery(qBuilder, filter, entityNameServer, modelRegistry);
      };

      expect(fn).toThrowErrorMatchingSnapshot();
    });

    it('reject if AND is not provided with an array', () => {
      const qBuilder = connection.createQueryBuilder(
        storageTableNameServer,
        entityNameServer,
      );

      const filter = {
        $and: {
          name: {
            $unknownOp: 123,
          },
        },
      };

      const fn = () => {
        buildWhereQuery(qBuilder, filter, entityNameServer, modelRegistry);
      };

      expect(fn).toThrowErrorMatchingSnapshot();
    });

    it('reject if OR is not provided with an array', () => {
      const qBuilder = connection.createQueryBuilder(
        storageTableNameServer,
        entityNameServer,
      );

      const filter = {
        $or: 123,
      };

      const fn = () => {
        buildWhereQuery(qBuilder, filter, entityNameServer, modelRegistry);
      };

      expect(fn).toThrowErrorMatchingSnapshot();
    });

    it('reject if operator is used without an attribute', () => {
      const qBuilder = connection.createQueryBuilder(
        storageTableNameServer,
        entityNameServer,
      );

      const filter = {
        $and: [
          {
            $lt: 123,
          },
        ],
      };

      const fn = () => {
        buildWhereQuery(qBuilder, filter, entityNameServer, modelRegistry);
      };

      expect(fn).toThrowErrorMatchingSnapshot();
    });

    it('reject if $not operator is used without a filter', () => {
      const qBuilder = connection.createQueryBuilder(
        storageTableNameServer,
        entityNameServer,
      );

      const filter = {
        $and: [
          {
            $not: 123,
          },
        ],
      };

      const fn = () => {
        buildWhereQuery(qBuilder, filter, entityNameServer, modelRegistry);
      };

      expect(fn).toThrowErrorMatchingSnapshot();
    });

    it('treat $in operator with empty value list as mistake and return no results', () => {
      const qBuilder = connection.createQueryBuilder(
        storageTableNameServer,
        entityNameServer,
      );

      const filter = {
        $and: [
          {
            ip: {
              $in: [],
            },
          },
        ],
      };

      buildWhereQuery(qBuilder, filter, entityNameServer, modelRegistry);
      const query = qBuilder.getQueryAndParameters();
      expect(query).toMatchSnapshot();
    });

    it('treat $notIn operator with empty value list as mistake and return no results', () => {
      const qBuilder = connection.createQueryBuilder(
        storageTableNameServer,
        entityNameServer,
      );

      const filter = {
        $and: [
          {
            ip: {
              $notIn: [],
            },
          },
        ],
      };

      buildWhereQuery(qBuilder, filter, entityNameServer, modelRegistry);
      const query = qBuilder.getQueryAndParameters();
      expect(query).toMatchSnapshot();
    });

    it('complex and nested filter with various operators', () => {
      const qBuilder = connection.createQueryBuilder(
        storageTableNameServer,
        entityNameServer,
      );

      const filter = {
        $or: [
          {
            ip: {
              $in: ['127.0.0.1', '192.168.0.1'],
            },
            name: {
              $startsWith: 'test-',
            },
          },
          {},
          {
            ip: '10.0.0.1',
          },
          {
            name: {
              $contains: 'open',
              $startsWith: 'demo',
              $endsWith: 'server',
            },
            $or: [
              {
                clusterZone: {
                  $in: [6, 7],
                },
              },
              {
                clusterZone: {
                  $gt: 10,
                  $lt: 100,
                  $ne: 50,
                  $notIn: [44],
                },
              },
            ],
            $and: [
              {
                $or: [
                  {
                    clusterZone: 88,
                  },
                  {
                    clusterZone: {
                      $lte: 40,
                      $gte: 30,
                    },
                  },
                  {
                    clusterZone: {
                      $noResult: true,
                    },
                  },
                ],
              },
              {
                name: {
                  $notContains: 'open',
                  $notStartsWith: 'demo',
                  $notEndsWith: 'server',
                },
              },
            ],
          },
        ],
        $and: [
          {
            name: 'secret-server',
          },
          {},
        ],
      };

      buildWhereQuery(qBuilder, filter, entityNameServer, modelRegistry);
      const query = qBuilder.getQueryAndParameters();
      expect(query).toMatchSnapshot();
    });

    it('sub query with complex filter', () => {
      const qBuilder = connection.createQueryBuilder(
        storageTableNameServer,
        entityNameServer,
      );

      const filter = {
        $and: [
          {
            ip: {
              $in: ['127.0.0.1', '192.168.0.1'],
            },
            name: {
              $startsWith: 'test-',
            },
          },
          {
            $sub: {
              entity: 'ClusterZone',
              condition: [
                {
                  targetAttribute: 'id',
                  operator: '$eq',
                  sourceAttribute: 'clusterZone',
                },
                {
                  targetAttribute: 'ip',
                  operator: '$in',
                  value: ['127.0.0.1', '192.168.0.1'],
                },
              ],
            },
          },
        ],
      };

      buildWhereQuery(qBuilder, filter, entityNameServer, modelRegistry, false);
      const query1 = qBuilder.getQueryAndParameters();
      expect(query1).toMatchSnapshot('without isGetMany');

      buildWhereQuery(qBuilder, filter, entityNameServer, modelRegistry, true);
      const query2 = qBuilder.getQueryAndParameters();
      expect(query2).toMatchSnapshot('with isGetMany');
    });
  });
});
