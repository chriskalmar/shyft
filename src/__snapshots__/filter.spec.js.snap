// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter buildWhereQuery combined logical AND/OR filter 1`] = `
Array [
  "SELECT \\"Server\\".\\"id\\" AS \\"Server_id\\", \\"Server\\".\\"cluster_zone\\" AS \\"Server_cluster_zone\\", \\"Server\\".\\"name\\" AS \\"Server_name\\", \\"Server\\".\\"ip\\" AS \\"Server_ip\\" FROM \\"server\\" \\"Server\\" WHERE ((ip = ? AND ((cluster_zone = ?) OR (cluster_zone = ?))) AND (name = ?))",
  Array [
    "127.0.0.1",
    1,
    7,
    "test-server-1",
  ],
]
`;

exports[`filter buildWhereQuery combined logical AND/OR filter 2`] = `
Array [
  "SELECT \\"Server\\".\\"id\\" AS \\"Server_id\\", \\"Server\\".\\"cluster_zone\\" AS \\"Server_cluster_zone\\", \\"Server\\".\\"name\\" AS \\"Server_name\\", \\"Server\\".\\"ip\\" AS \\"Server_ip\\" FROM \\"server\\" \\"Server\\" WHERE (((ip = ? AND (cluster_zone = ?) AND (name = ?)) OR (name = ?)))",
  Array [
    "127.0.0.1",
    1,
    "test-server-110",
    "test-server-1",
  ],
]
`;

exports[`filter buildWhereQuery complex and nested filter with various operators 1`] = `
Array [
  "SELECT \\"Server\\".\\"id\\" AS \\"Server_id\\", \\"Server\\".\\"cluster_zone\\" AS \\"Server_cluster_zone\\", \\"Server\\".\\"name\\" AS \\"Server_name\\", \\"Server\\".\\"ip\\" AS \\"Server_ip\\" FROM \\"server\\" \\"Server\\" WHERE (((ip IN (?, ?) AND name ILIKE ?) OR (ip = ?) OR (name ILIKE ? AND name ILIKE ? AND name ILIKE ? AND ((cluster_zone IN (?, ?)) OR (cluster_zone > ? AND cluster_zone < ? AND cluster_zone <> ? AND cluster_zone NOT IN (?))) AND (((cluster_zone = ?) OR (cluster_zone <= ? AND cluster_zone >= ?) OR (TRUE IS FALSE))) AND (name NOT ILIKE ? AND name NOT ILIKE ? AND name NOT ILIKE ?))) AND (name = ?))",
  Array [
    "127.0.0.1",
    "192.168.0.1",
    "test-%",
    "10.0.0.1",
    "%open%",
    "demo%",
    "%server",
    6,
    7,
    10,
    100,
    50,
    44,
    88,
    40,
    30,
    "%open%",
    "demo%",
    "%server",
    "secret-server",
  ],
]
`;

exports[`filter buildWhereQuery empty filter 1`] = `
Array [
  "SELECT \\"Server\\".\\"id\\" AS \\"Server_id\\", \\"Server\\".\\"cluster_zone\\" AS \\"Server_cluster_zone\\", \\"Server\\".\\"name\\" AS \\"Server_name\\", \\"Server\\".\\"ip\\" AS \\"Server_ip\\" FROM \\"server\\" \\"Server\\"",
  Array [],
]
`;

exports[`filter buildWhereQuery logical AND filter 1`] = `
Array [
  "SELECT \\"Server\\".\\"id\\" AS \\"Server_id\\", \\"Server\\".\\"cluster_zone\\" AS \\"Server_cluster_zone\\", \\"Server\\".\\"name\\" AS \\"Server_name\\", \\"Server\\".\\"ip\\" AS \\"Server_ip\\" FROM \\"server\\" \\"Server\\" WHERE ((ip = ? AND name = ?) AND (cluster_zone = ?))",
  Array [
    "127.0.0.1",
    "test-server-1",
    4,
  ],
]
`;

exports[`filter buildWhereQuery logical OR filter 1`] = `
Array [
  "SELECT \\"Server\\".\\"id\\" AS \\"Server_id\\", \\"Server\\".\\"cluster_zone\\" AS \\"Server_cluster_zone\\", \\"Server\\".\\"name\\" AS \\"Server_name\\", \\"Server\\".\\"ip\\" AS \\"Server_ip\\" FROM \\"server\\" \\"Server\\" WHERE (((ip = ? AND name = ?) OR (ip = ? AND name = ?)))",
  Array [
    "127.0.0.1",
    "test-server-1",
    "127.0.0.1",
    "test-server-2",
  ],
]
`;

exports[`filter buildWhereQuery reject if $not operator is used without a filter 1`] = `"$not expects a filter"`;

exports[`filter buildWhereQuery reject if AND is not provided with an array 1`] = `"$and expects an array of filters"`;

exports[`filter buildWhereQuery reject if OR is not provided with an array 1`] = `"$or expects an array of filters"`;

exports[`filter buildWhereQuery reject if operator is used without an attribute 1`] = `"missing attribute name for filter operator"`;

exports[`filter buildWhereQuery reject if unknown operators are used 1`] = `"Unknown filter operator: $unknownOp"`;

exports[`filter buildWhereQuery simple properties filter 1`] = `
Array [
  "SELECT \\"Server\\".\\"id\\" AS \\"Server_id\\", \\"Server\\".\\"cluster_zone\\" AS \\"Server_cluster_zone\\", \\"Server\\".\\"name\\" AS \\"Server_name\\", \\"Server\\".\\"ip\\" AS \\"Server_ip\\" FROM \\"server\\" \\"Server\\" WHERE (ip = ? AND name = ?)",
  Array [
    "127.0.0.1",
    "test-server-1",
  ],
]
`;

exports[`filter purifyFilter should clean empty filter bocks 1`] = `
Object {
  "$and": Array [
    Object {
      "$or": Array [
        Object {
          "LookupEntity.role": 1,
        },
        Object {
          "accessLevel": 1,
        },
      ],
      "language": Object {
        "$gte": 1,
        "$lt": 5,
      },
      "username": Object {
        "$contains": "bot",
        "$startsWith": "_system_%",
      },
    },
    Object {
      "$or": Array [
        Object {
          "$and": Array [
            Object {
              "username": "hallo",
            },
          ],
        },
      ],
    },
  ],
  "$or": Array [
    Object {
      "lastname": "smith",
    },
  ],
}
`;
